OCPBUGS-18649:  E2E: PPC Test cases

mrniranjan opened the pull request:
This test cases uses podman command to test PPC. 

Existing PPC automation uses Compiled PPC binary to execute the tests, These tests cannot be executed on Downstream since PPC binary is part of the NTO image which can be called only using podman command. 

This PR uses gexec library to execute Podman cli and test all PPC Command line options.

ffromani commented on test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:
```diff
@@ -0,0 +1,58 @@
+package __performance_ppc
```
we should probably rename all the other suites to read like `00_config` and so forth, and this should probably be done in its own prerequisite PR

ffromani commented on test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+const (
+	nto_downstream_image_registry = "registry-proxy.engineering.redhat.com/rh-osbs/openshift-ose-cluster-node-tuning-operator@sha256"
```
no internal references in public code please

mrniranjan replied:
ack, fixed in the latest commit

ffromani commented on test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:
```diff
@@ -19,13 +19,19 @@ const (
 	var err error
 	var nto_image_registry = ""
+	By("Check podman binary exists")
+	_, err = os.Stat(podmanBinary)
+	if err != nil {
+		Skip(fmt.Sprintf("%s doesn't exists", podmanBinary))
+	}
```
I'd use https://pkg.go.dev/os/exec#LookPath

ffromani reviewed (commented):
I for myself would not add an exploded must-gather tree, but a .tar.gz archive with the same content, to be extracted (and of course cleaned up) just before to run the actual test.
It's true we add a blob to the repo, but I think it's actually less noise than adding so many files. Perhaps we can abuse a bit quay.io and add data image(s) with the archive(s) we need, download and unpack that image as part as the testing process. This is more complicated but leaves this repo the cleanest.

mrniranjan commented:
> I for myself would not add an exploded must-gather tree, but a .tar.gz archive with the same content, to be extracted (and of course cleaned up) just before to run the actual test. It's true we add a blob to the repo, but I think it's actually less noise than adding so many files. Perhaps we can abuse a bit quay.io and add data image(s) with the archive(s) we need, download and unpack that image as part as the testing process. This is more complicated but leaves this repo the cleanest.

Thanks, I have removed the mustgather directory and added MUSTGATHER_DIR environment variable to provided pointing to appropriate directory .

mrniranjan commented:
/retest

mrniranjan commented:
/retest

Tal-or commented:
@mrniranjan the test has been skipped and didn't actually run:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-node-tuning-operator/708/pull-ci-openshift-cluster-node-tuning-operator-master-e2e-gcp-pao/1691245834059911168/artifacts/e2e-gcp-pao/test/build-log.txt

mrniranjan commented:
> @mrniranjan the test has been skipped and didn't actually run: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-node-tuning-operator/708/pull-ci-openshift-cluster-node-tuning-operator-master-e2e-gcp-pao/1691245834059911168/artifacts/e2e-gcp-pao/test/build-log.txt
> 
> ```
> Will run �[1m4�[0m of �[1m4�[0m specs
> �[38;5;243m------------------------------�[0m
> �[0m�[1m[BeforeSuite] �[0m
> �[38;5;243m/go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:22�[0m
>   �[1mSTEP:�[0m Check podman binary exists �[38;5;243m@ 08/15/23 01:32:58.128�[0m
>   �[38;5;14m[SKIPPED]�[0m in [BeforeSuite] - /go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:27 �[38;5;243m@ 08/15/23 01:32:58.128�[0m
> �[38;5;14mS [SKIPPED] [0.000 seconds]�[0m
> �[38;5;14m�[1m[BeforeSuite] �[0m
> �[38;5;243m/go/src/github.com/openshift/cluster-node-tuning-operator/test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:22�[0m
> 
>   �[38;5;14m[SKIPPED] /usr/bin/podman doesn't exists�[0m
> ```

@Tal-or These tests will not run in github pr ci because not only does it doesn't contain the podman binary and also because the these tests required mustgather  data (from 4.14). So these test will only run in downstream

Tal-or commented on test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:
```diff
@@ -0,0 +1,42 @@
+
+})
+
+func Test9Ppc(t *testing.T) {
```
Maybe using the full name and remove the `9` is more readable

mrniranjan replied:
fixed in the latest commit

mrniranjan commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,205 @@
+	PodmanBinary      string
+}
+
+type PPCSession struct {
```
Because we want to extend this to add new methods under ppc integration in future, we want to port some of the existing ppc tests using this method.

mrniranjan commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,205 @@
+			Expect(*pp.Spec.CPU.Reserved).To(Equal(v2.CPUSet("0,2")))
+			Expect(*pp.Spec.WorkloadHints.RealTime).To(BeTrue())
+			Expect(*pp.Spec.NUMA.TopologyPolicy).To(Equal("restricted"))
+			Eventually(session).Should(gexec.Exit(0))
```
Yes its valid, because the session variable even though it's a struct variable , the underlying type *gexec.session has the gomegaMatcher

mrniranjan commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,205 @@
+			errString := "Error: failed to compute the reserved and isolated CPUs: please specify the reserved CPU count in the range [1,3]"
+			ok, err := regexp.MatchString(errString, string(output))
+			if err != nil {
+				testlog.Error(err.Error())
```
fixed in the latest commit

mrniranjan commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,205 @@
+			errString := "Error: failed to compute the reserved and isolated CPUs: can't allocate odd number of CPUs from a NUMA Node"
+			ok, err := regexp.MatchString(errString, string(output))
+			if err != nil {
+				testlog.Error(err.Error())
```
fixed in the latest commit

shajmakh commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,200 @@
+	return &PPCSession{session}, nil
+}
+
+type PPCTestIntegration struct {
+	PPCTest
+}
+
+type PPCSessionIntegration struct {
+	*PPCSession
+}
```
why are we redefining another struct with single field (which is a struct too)? do we plan to add more fields to this?

mrniranjan replied:
Removed in the latest commit.

shajmakh commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,200 @@
+	*PPCSession
+}
+
+func (p *PPCTestIntegration) MakeOptions(args []string) []string {
+	return args
+}
```
if this returns the same args we passed, then why do we need a method? is it for readability (just guessing) only?

mrniranjan replied:
Currently it may seem to be doing nothing, but i want to extend this in future, for example some of the default options i can put it here and in the actual test, i can pass the actual options that are required for test.  Also i can use this function exclusively for all the podman options.

shajmakh commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,200 @@
+			Expect(err).ToNot(HaveOccurred(), "Unable to marshal the ppc output")
+			cpuset := *pp.Spec.CPU.Reserved
+			totalReservedCpus := len(strings.Split(string(cpuset), ","))
+			Expect(totalReservedCpus == reservedCpuCount).To(BeTrue())
```
you can use gomega To(Equal(..)) instead: 
Expect(ACTUAL).To(Equal(EXPECTED))

mrniranjan replied:
fixed in the latest commit

shajmakh commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,200 @@
+			err = yaml.Unmarshal(output, pp)
+			Expect(err).ToNot(HaveOccurred(), "Unable to marshal the ppc output")
+			cpuset := *pp.Spec.CPU.Reserved
+			totalReservedCpus := len(strings.Split(string(cpuset), ","))
```
I believe you can use Size of the cpuset instead

mrniranjan replied:
fixed in the latest commit

shajmakh commented on test/e2e/performanceprofile/functests/10_performance_ppc/ppc.go:
```diff
@@ -0,0 +1,200 @@
+				fmt.Sprintf("%s:%s:z", mustgatherDir, mustgatherDir),
+				ntoImage,
+				"--mcp-name=worker",
+				fmt.Sprintf("--reserved-cpu-count=%d", 100),
```
I know we don't usually have systems with CPU amount greater than 80, but it would be more certain to set this filed based on the actual CPU count on the system that runs this test, for example get the CPU count and add 10

mrniranjan replied:
Majority of test is reading mustgather directory and uses that to generate Profile, the only time we use the cluster is to verify if the generated Performance Profile  is valid, So it doesn't matter where this test is executed. I want to make the  tests agnostic to cluster and only required is mustgather and a connection to cluster to verify if generated PP is valid. 

I can increase the number to another 100 or so

shajmakh replied:
got it, thanks for clarifying

shajmakh reviewed (commented):
Thanks for the PR, left a few comments below.

mrniranjan commented on Makefile:
```diff
@@ -201,7 +201,7 @@ pao-functests: cluster-label-worker-cnf pao-functests-only
 	@echo "Cluster Version"
 	hack/show-cluster-version.sh
-	hack/run-test.sh -t "test/e2e/performanceprofile/functests/0_config test/e2e/performanceprofile/functests/1_performance test/e2e/performanceprofile/functests/6_mustgather_testing" -p "-v -r --fail-fast  --flake-attempts=2 --junit-report=report.xml" -m "Running Functional Tests"
+	hack/run-test.sh -t "test/e2e/performanceprofile/functests/0_config test/e2e/performanceprofile/functests/1_performance test/e2e/performanceprofile/functests/6_mustgather_testing test/e2e/performanceprofile/functests/10_performance_ppc" -p "-v -r --fail-fast  --flake-attempts=2 --junit-report=report.xml" -m "Running Functional Tests"
```
the actual ppc tests do not take more than 5 minutes

shajmakh replied:
do they trigger system reboot?

shajmakh replied:
I mixed PRs, please ignore my comment about the reboot

mrniranjan commented on test/e2e/performanceprofile/functests/10_performance_ppc/10_ppc_suite_test.go:
```diff
@@ -0,0 +1,39 @@
+	By("Checking Environment Variables")
+	testlog.Infof("NTO Image used: %s", testutils.NtoImageRegistry)
+	if testutils.MustGatherDir == "" {
+		Skip("set env variable MUSTGATHER_DIR to ocp mustgather directory")
```
hmm, i don't want to use default path because we run the tests on different environment, disconnected cluster, local laptop, clusterbot, downstream ci and each ci runs mustgather in a different fashion , where the mustgather is extracted in different location .

If we use an environment variable, we can modify the CI to use environment variable pointing to the extracted mustgather directory.

shajmakh replied:
ok makes sense, thanks for clarifying.

mrniranjan commented on test/e2e/performanceprofile/functests/utils/consts.go:
```diff
@@ -24,6 +24,12 @@ var NodesSelector string
 // ProfileNotFound is true when discovery mode is enabled and no valid profile was found
 var ProfileNotFound bool
 
+// NtoImageRegistry represents NTO Image registry which is either quay.io or any other internal registry
+var NtoImageRegistry string
```
image is located in some registry. either quay.io or internal mirror or internal registry.

mrniranjan replied:
I can see the confusion, fixing it

mrniranjan replied:
Renamed it in the latest commit

mrniranjan commented:
/jira refresh

ffromani commented:
/approve
/lgtm
/hold

ffromani commented:
/approve

ffromani commented:
/approve

ffromani commented:
/lgtm

ffromani commented:
/hold cancel

we reached consensus offline to move forward

ffromani commented:
/retest

mrniranjan commented:
/retest-required

mrniranjan commented:
/retest-required
